### PR TITLE
Implement NebulexLocalDistributedAdapter

### PR DESCRIPTION
**Pull request by martosaur**

### CODE REVIEW

File: lib/nebulex_local_distributed_adapter.ex [hunk 1]
```diff
@@ -1,18 +1,225 @@
...
+  def init(opts) do
+    opts = Keyword.put(opts, :model, :inclusive)
+
+    {:ok, child_spec, adapter_meta} = Nebulex.Adapters.Multilevel.init(opts)
+
+    [l1_meta | _] = adapter_meta.levels
+
+    if l1_meta.cache.__adapter__ != Nebulex.Adapters.Local do
+      raise "L1 must be Nebulex.Adapters.Local"
+    end
```
Comment (jhchen) [inline lib/nebulex_local_distributed_adapter.ex]: The user doesn't have a choice except to use Nebulex.Adapters.Local so why give them a "choice"? We should just create/manage the L1 local adapter for the user so they can't mess this up. We could accept optional parameters to pass to the creation but be explicit this is the purpose ex `:primary` like [Replicated](https://hexdocs.pm/nebulex/Nebulex.Adapters.Replicated.html#module-options) does although I would call it `:local` or something similar

File: lib/nebulex_local_distributed_adapter.ex [hunk 2]
```diff
@@ -1,18 +1,225 @@
...
+  1. L1 must always be `Nebulex.Adapters.Local`
+  2. Other levels must be global for nodes, meaning they behave like a shared
+    storage. Simplest example is `NebulexRedisAdapter`, but `Replicated` and
+    `Partitioned` should work too.
+  3. All write operations are asynchronously broadcasted to other nodes which
+    invalidate affected keys in their local L1 caches.
 
+  > #### Race conditions {: .warning}
+  > Due to asynchronous nature of invalidation it is possible that a node will read
+  stale value from its local cache.
```
Comment (jhchen) [inline lib/nebulex_local_distributed_adapter.ex]: Not sure why this is called out as a warning. I don't think users will expect all nodes to instantly get the new value so eventual consistency is expected.

One thing I do think we should warn users about is to delete/invalidate data from higher levels than this cache (ex Ecto database) first, then immediately calling this cache's invalidation. Otherwise we do run into stale values possibly persisting as similarly described in https://github.com/slab/nebulex_clustered_multilevel_adapter/pull/1#issuecomment-1288260150

Also I think we should find a way to not duplicate this content from the README. Here seems more important than the README since the hexdocs site is generated from here.

**Review by jhchen — CHANGES_REQUESTED** [review]

File: lib/nebulex_local_distributed_adapter.ex (hunk 1 above)
Comment (martosaur) [reply lib/nebulex_local_distributed_adapter.ex]: I think my initial reasoning was that keeping configuration a full copy of `Multilevel` would just give a better understanding of what the adapter actually is. i.e. this
```elixir
defmodule MyCache do
  use Nebulex.Cache,
    otp_app: :nebulex_local_distributed_adapter,
    adapter: NebulexLocalDistributedAdapter,
    levels: [
      {MyCache.Local, [name: Foo]},
      {MyCache.Redis, [name: Bar]}
    ]
end
```

IMO is more familiar and thus straightforward than this:
```elixir
defmodule MyCache do
  use Nebulex.Cache,
    otp_app: :nebulex_local_distributed_adapter,
    adapter: NebulexLocalDistributedAdapter,
    local_opts: [name: Foo],
    levels: [
      {MyCache.Redis, [name: Bar]}
    ]
end

# or

defmodule MyCache do
  use Nebulex.Cache,
    otp_app: :nebulex_local_distributed_adapter,
    adapter: NebulexLocalDistributedAdapter,
    local: {MyCache.Local, [name: Foo]},
    levels: [
      {MyCache.Redis, [name: Bar]}
    ]
end
```

Anyway, not much objection from my side for creating local adapter within init, so I'll just implement it.

File: lib/nebulex_local_distributed_adapter.ex (hunk 2 above)
Comment (martosaur) [reply lib/nebulex_local_distributed_adapter.ex]: > One thing I do think we should warn users about is to delete/invalidate data from higher levels than this cache (ex Ecto database) first, then immediately calling this cache's invalidation. Otherwise we do run into stale values possibly persisting as similarly described in https://github.com/slab/nebulex_clustered_multilevel_adapter/pull/1#issuecomment-1288260150

Isn't this the basic usage of any cache? Invalidation must happen after change.

Comment (martosaur) [reply lib/nebulex_local_distributed_adapter.ex]: > Not sure why this is called out as a warning. I don't think users will expect all nodes to instantly get the new value so eventual consistency is expected.

They might expect that invalidation is guaranteed to complete by the time the function returns. That would be the case if we used multicall (vs multicast).

Comment (jhchen) [reply lib/nebulex_local_distributed_adapter.ex]: It sounds like we are both concerned about the users' assumptions about async/blocking but disagree on the implications.

> Invalidation must happen after change.

If the user thinks the cache invalidate and database delete calls are blocking, the order would not matter so they might happen to do it in the wrong order. We do know Ecto delete/update is blocking and even if it isn't or they are using something else, our warning is saying "make sure it's deleted from the primary location before invalidating" so it's not really our concern. But it does give more weight to the user assuming the order would not matter. I would not suggest this warning if the audience were just Slab engineers but this is to be open sourced for the public. But don't feel strongly.

> They might expect that invalidation is guaranteed to complete by the time the function returns. That would be the case if we used multicall (vs multicast).

If you're saying specifically that this is possible:

"stale" = LocalDistributed.get("key")
LocalDistributed.delete("key")
"stale" = LocalDistributed.get("key")

I don't believe this is the case given the [current implementation](https://github.com/slab/nebulex_local_distributed_adapter/pull/1/files#diff-85b372aa64a4d1fa13a600827a0b127f58126b5414196606c506c8750776aa2cR94). We delete locally and to Redis first and they are blocking (I tested with [this](https://github.com/slab/slab#simulating-network-latency)). The async part is invalidating on remote servers. Nebulex in general is blocking so the different `levels` configurations should not change this (if it were not then there should probably be a warning in the Nebulex repo).

Comment (martosaur) [reply lib/nebulex_local_distributed_adapter.ex]: > If the user thinks the cache invalidate and database delete calls are blocking, the order would not matter so they might happen to do it in the wrong order.

The order always matters if there is any concurrency involved. Even in single node environment with both calls blocking this can happen:
1. Client A invalidates cache
2. Client B reads from DB and writes to cache
3. Client A updates DB.

Am I missing something here?

> If you're saying specifically that this is possible:
"stale" = LocalDistributed.get("key")
LocalDistributed.delete("key")
"stale" = LocalDistributed.get("key")

A slightly different variation is possible:
1. node: A -> "stale" = LocalDistributed.get("key")
2. node A -> LocalDistributed.delete("key")
3. node B -> "stale" = LocalDistributed.get("key")
Basically, yes, it falls under eventual consistency definition, but the degree of "eventuality" is different depending on cast vs call implementation and I thought it's worth mentioning.

Comment (jhchen) [reply lib/nebulex_local_distributed_adapter.ex]: > The order always matters if there is any concurrency involved... Am I missing something here?

My example was only considering deletes when I said the order of those deletes didn't matter if they are blocking. I chose to discuss delete because it's the simplest case. Yes ordering is important for concurrency and we're not here in a adapter library repo to teach people about concurrency fundamentals. But the nebulex main repo ordered incorrectly which led to https://github.com/cabol/nebulex/pull/174 so I felt the wider audience of nebulex users could benefit when contributors made that error.

> A slightly different variation is possible...

Even if `delete` is blocking in step 2, node B can still get "stale" in step 3. It'd have to be blocking or isolating (ex locks) for the entire cluster. This is not an expectation I'd hold which is why I thought it odd to call out and warn against.

Again don't feel strongly on either but that is why.

Comment (martosaur) [reply lib/nebulex_local_distributed_adapter.ex]: Alright, I rephrased to squeeze both warnings in there.

**Review by jhchen — APPROVED** [review]
